server/server.cpp

```diff
@@ -0,0 +1,136 @@
+    buttonLayout-> addStretch(1);
+    QVBoxLayout * mainLayout = nullptr;
+    if (QGuiApplication::styleHints()->showIsFullScreen() || QGuiApplication::styleHints()->showIsMaximized()) {
+        auto outerVerticalLayout = new QVBoxLayout(this);
+        outerVerticalLayout->addItem(new QSpacerItem(0, 0, QSizePolicy::Ignored, QSizePolicy::MinimumExpanding));
+        auto outerHorizontalLayout = new QHBoxLayout;
+        outerHorizontalLayout->addItem(new QSpacerItem(0, 0, QSizePolicy::MinimumExpanding, QSizePolicy::Ignored));
+        auto groupBox = new QGroupBox(QGuiApplication::applicationDisplayName());
+        mainLayout = new QVBoxLayout(groupBox);
+        outerHorizontalLayout->addWidget(groupBox);
+        outerHorizontalLayout->addItem(new QSpacerItem(0, 0, QSizePolicy::MinimumExpanding, QSizePolicy::Ignored));
+        outerVerticalLayout->addLayout(outerHorizontalLayout);
+        outerVerticalLayout->addItem(new QSpacerItem(0, 0, QSizePolicy::Ignored, QSizePolicy::MinimumExpanding));
+    } else {
+        mainLayout = new QVBoxLayout(this);
+    }
+
+    mainLayout->addWidget(statusLabel);
+    mainLayout->addLayout(buttonLayout);
+
+    setWindowTitle(QGuiApplication::applicationDisplayName());
+}
+
+void Server::initServer()
+{
+    tcpServer = new QTcpServer(this);
+    if(!tcpServer->listen())
+    {
+        QMessageBox::critical (this,tr("Server"),tr("unable to start the server: %1.").arg(tcpServer->errorString()));
+        close();
+        return;
+
+    }
+    QString ipAddress;
+    QList<QHostAddress> ipAddressesList = QNetworkInterface::allAddresses();
+    for (int i = 0; i <ipAddressesList.size();++i)
+    {
+        if (ipAddressesList.at(i) != QHostAddress::LocalHost &&
+                ipAddressesList.at(i).toIPv4Address())
+        {
+            ipAddress = ipAddressesList.at(i).toString();
+            break;
+        }
+    }
+    if (ipAddress.isEmpty())
+        ipAddress = QHostAddress(QHostAddress::LocalHost).toString();
+    statusLabel->setText(tr("The server is running on\n\nIP: %1\nport: %2\n\n"
+                            "Run Client now.")
+                         .arg(ipAddress).arg(tcpServer->serverPort()));
+
+
+}
+
+void Server::send()
+{
+
+    QByteArray block ;
+    QDataStream out(&block, QIODevice::WriteOnly);
+    out.setVersion(QDataStream::Qt_5_10);
+
+    out << list;
+    qDebug() << block << block.size();
+
+    QTcpSocket *clientConnection = tcpServer->nextPendingConnection();
+    connect(clientConnection, &QAbstractSocket::disconnected,
+            clientConnection, &QObject::deleteLater);
+
+    clientConnection->write(block);
+    clientConnection->disconnectFromHost();
+}
+
+QStringList Server::filereader()
+{
+
+    QFile myFile ("C:\\Users\\user\\Desktop\\Urlfile.txt");
+
+
+    if(!myFile.open(QIODevice::OpenModeFlag::ReadOnly))
+    {
+        qCritical() << "can not open file";
+        qCritical() << myFile.errorString();
+
+
+    }
+    QTextStream stream(& myFile);
```

Comment:
bad case: file opening is failed, but you still put the stream on it... You need to stop further processing if the file opening is not successful.